### ISSUE TITLE
Properly detect and report an unreachable guest

### ISSUE THREAD

**Issue by RHEmployee** [teemtee/tmt#1121]
`tmt -vvvddd run provision -h connect -g 123`

After:
`err: ssh: connect to host 0.0.0.123 port 22: Connection timed out`

tmt just ignores error for timeout and count this as non presence of  `/etc/os-release`
and in next move, tmt tries to check existence of `/etc/lsb-release`, then `/etc/redhat-release`, etc ...

This doesn't make much sense while machine can be unreachable (still in phase of booting for example).

```
       key: not provided
            Using private key authentication.
            Execute command 'cat /etc/os-release' on guest '123'.
            Create the master ssh connection: ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -S /run/user/20787/tmt/tmplr5yq9jn -MNnT root@123
            Run command 'ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -S /run/user/20787/tmt/tmplr5yq9jn root@123 export TMT_PLAN_DATA=/var/tmp/tmt/run-003/plans/features/advanced/data; cat /etc/os-release'.
            err: ssh: connect to host 0.0.0.123 port 22: Connection timed out
            Command returned '255'.
            Execute command 'cat /etc/lsb-release' on guest '123'.
            Run command 'ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -S /run/user/20787/tmt/tmplr5yq9jn root@123 export TMT_PLAN_DATA=/var/tmp/tmt/run-003/plans/features/advanced/data; cat /etc/lsb-release'.
```

**Comment by RHEmployee** [comment]
After checking how ssh cli works, I don't think there is easy way how to handle timeout error by return code. Ssh cli returns 255 even on usage so only way how to handle specific error (timeout in this case) is to parse output of ssh.

Do you think that we should parse output of ssh in case of error return codes?

**Comment by psss** [comment]
> tmt just ignores error for timeout and count this as non presence of `/etc/os-release`

Yes, this should be handled properly. And seems that it is working as expected (almost) in the `virtual` plugin:

    tmt run provision --how virtual --user foo login finish

Repeatedly tries to reconnect and run `whoami` until the timeout:

    fail: Failed to connect in 120s.

But! Then continues with `cat /etc/os-release` and further detection commands until:

    Package manager was not detected on guest 'default-0'.

For some reason, the `connect` plugin does not even retry the `whoami` and attempts one unsuccessfull command after another:

    Spawning the SSH master process: ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oConnectionAttempts=5 -oConnectTimeout=60 -oServerAliveInterval=5 -oServerAliveCountMax=60 -p10042 -oPasswordAuthentication=no -S/var/tmp/tmt/run-297/ssh-sockets/127.0.0.1-10042-root.socket -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oConnectionAttempts=5 -oConnectTimeout=60 -oServerAliveInterval=5 -oServerAliveCountMax=60 -p10042 -oPasswordAuthentication=no -S/var/tmp/tmt/run-297/ssh-sockets/127.0.0.1-10042-root.socket -MNnT root@127.0.0.1
    Execute command 'export TMT_VERSION=1.53.0.dev8+gc41959ec; export TMT_TREE=/var/tmp/tmt/run-297/default/plan/tree; export TMT_PLAN_DATA=/var/tmp/tmt/run-297/default/plan/data; export TMT_PLAN_ENVIRONMENT_FILE=/var/tmp/tmt/run-297/default/plan/data/variables.env; cat /etc/os-release' on guest '127.0.0.1'.
    Workdir '/var/tmp/tmt/run-297/default/plan/provision/default-0' created.
    Run command: ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oConnectionAttempts=5 -oConnectTimeout=60 -oServerAliveInterval=5 -oServerAliveCountMax=60 -p10042 -oPasswordAuthentication=no -S/var/tmp/tmt/run-297/ssh-sockets/127.0.0.1-10042-root.socket root@127.0.0.1 'export TMT_VERSION=1.53.0.dev8+gc41959ec; export TMT_TREE=/var/tmp/tmt/run-297/default/plan/tree; export TMT_PLAN_DATA=/var/tmp/tmt/run-297/default/plan/data; export TMT_PLAN_ENVIRONMENT_FILE=/var/tmp/tmt/run-297/default/plan/data/variables.env; cat /etc/os-release'
    environment
    err: Warning: Permanently added '[127.0.0.1]:10042' (ED25519) to the list of known hosts.
    err: root@127.0.0.1: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).
    Command returned '255' (unrecognized).
    Execute command 'export TMT_VERSION=1.53.0.dev8+gc41959ec; export TMT_TREE=/var/tmp/tmt/run-297/default/plan/tree; export TMT_PLAN_DATA=/var/tmp/tmt/run-297/default/plan/data; export TMT_PLAN_ENVIRONMENT_FILE=/var/tmp/tmt/run-297/default/plan/data/variables.env; cat /etc/lsb-release' on guest '127.0.0.1'.
    Run command: ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oConnectionAttempts=5 -oConnectTimeout=60 -oServerAliveInterval=5 -oServerAliveCountMax=60 -p10042 -oPasswordAuthentication=no -S/var/tmp/tmt/run-297/ssh-sockets/127.0.0.1-10042-root.socket root@127.0.0.1 'export TMT_VERSION=1.53.0.dev8+gc41959ec; export TMT_TREE=/var/tmp/tmt/run-297/default/plan/tree; export TMT_PLAN_DATA=/var/tmp/tmt/run-297/default/plan/data; export TMT_PLAN_ENVIRONMENT_FILE=/var/tmp/tmt/run-297/default/plan/data/variables.env; cat /etc/lsb-release'
    environment
    err: Warning: Permanently added '[127.0.0.1]:10042' (ED25519) to the list of known hosts.
    err: root@127.0.0.1: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).
    Command returned '255' (unrecognized).
    Execute command 'export TMT_VERSION=1.53.0.dev8+gc41959ec; export TMT_TREE=/var/tmp/tmt/run-297/default/plan/tree; export TMT_PLAN_DATA=/var/tmp/tmt/run-297/default/plan/data; export TMT_PLAN_ENVIRONMENT_FILE=/var/tmp/tmt/run-297/default/plan/data/variables.env; arch' on guest '127.0.0.1'.
    ...

Reproducible easily:

    # terminal one
    tmt run provision login finish

    # terminal two (pick guest ip & port from the first one)
    tmt run -vvvddd provision -h connect --guest 127.0.0.1 --port 10042 login finish

We should bail out upon the first command when we detect that it's not possible to connect to the guest and give a reasonable error to the user so that they are not confused by irrelevant messages.

**Comment by psss** [comment]
Another user run into this. Adding to the backlog and raising priority. The fix should be most probably raising the right exception at the right time. Could possibly be one of the good first issues, fyi @vaibhavdaren & @AthreyVinay.

**Comment by tcerna** [comment]
Hi, I hit the same issue too when I tried to access an IP of unavailable machine.

```
   prepare
        queued push task #1: push to default-0
        push task #1: push to default-0
        cmd: rsync --version
        err: ssh: connect to host 192.168.187.44 port 22: Network is unreachable
    fail: Package manager was not detected on guest 'default-0'.
    finish
    
    Prune '/plans/newstyle/upgrades/tests/destructive_no_upgrade/test_preupgrade_not_inhibited_with_dev_kernels.py::TestPreupgradeNotInhibitedWithDevKernels' plan workdir '/var/tmp/tmt/run-102/plans/newstyle/upgrades/tests/destructive_no_upgrade/test_preupgrade_not_inhibited_with_dev_kernels-py-TestPreupgradeNotInhibitedWithDevKernels'.
        summary: 0 tasks completed

plan failed

The exception was caused by 1 earlier exceptions

Cause number 1:
    prepare step failed
    The exception was caused by 1 earlier exceptions
    Cause number 1:
        Package manager was not detected on guest 'default-0'.

```
